Practical 1: Sentiment Detection in Movie Reviews
========================================



This practical concerns detecting sentiment in movie reviews. This is a typical NLP classification task.
In [this file](https://gist.githubusercontent.com/bastings/d47423301cca214e3930061a5a75e177/raw/5113687382919e22b1f09ce71a8fecd1687a5760/reviews.json) (80MB) you will find 1000 positive and 1000 negative **movie reviews**.
Each review is a **document** and consists of one or more sentences.

To prepare yourself for this practical, you should
have a look at a few of these texts to understand the difficulties of
the task: how might one go about classifying the texts? You will write
code that decides whether a movie review conveys positive or
negative sentiment.

Please make sure you have read the following paper:

>   Bo Pang, Lillian Lee, and Shivakumar Vaithyanathan
(2002).
[Thumbs up? Sentiment Classification using Machine Learning
Techniques](https://dl.acm.org/citation.cfm?id=1118704). EMNLP.

Bo Pang et al. were the "inventors" of the movie review sentiment
classification task, and the above paper was one of the first papers on
the topic. The first version of your sentiment classifier will do
something similar to Pang et al.'s system. If you have questions about it,
you should resolve as soon as possible with your TA.


**Advice**

Please read through the entire practical and familiarise
yourself with all requirements before you start coding or otherwise
solving the tasks. Writing clean and concise code can make the difference
between solving the assignment in a matter of hours, and taking days to
run all experiments.

**Implementation**

While we inserted code cells to indicate where you should implement your own code, please feel free to add/remove code blocks where you see fit (but make sure that the general structure of the assignment is preserved). Also, please keep in mind that it is always good practice to structure your code properly, e.g., by implementing separate classes and functions that can be reused.

Generally, any function that we already use or import in the notebook can be used. Besides those, importing something like `deepcopy` or `tqdm` that does not change the implementation of the algorithms is fine. Functions that change or simplify the way you would implement the algorithm, including text processing functions from libraries like `sklearn`, `pandas` or `nltk` are not allowed unless specified (e.g. `ngrams` and `PorterStemmer` from `ntlk`). If you have questions about any specific function or library, please discuss this with your TA who will be the one grading your assignment.

## Environment

All code should be written in **Python 3**.
This is the default in Google Colab.

In [97]:
from IPython.utils import docs
from sklearn.metrics import confusion_matrix
!python --version


Python 3.12.4


If you want to run code on your own computer, then download this notebook through `File -> Download .ipynb`.
The easiest way to
install Python is through downloading
[Anaconda](https://www.anaconda.com/download).
After installation, you can start the notebook by typing `jupyter notebook filename.ipynb`.
You can also use an IDE
such as [PyCharm](https://www.jetbrains.com/pycharm/download/) to make
coding and debugging easier. It is good practice to create a [virtual
environment](https://docs.python.org/3/tutorial/venv.html) for this
project, so that any Python packages don’t interfere with other
projects.


**Learning Python 3**

If you are new to Python 3, you may want to check out a few of these resources:
- https://learnxinyminutes.com/docs/python3/
- https://www.learnpython.org/
- https://docs.python.org/3/tutorial/

In [20]:
import hashlib
import json
import math
import typing
from collections import Counter, defaultdict

import matplotlib.pyplot as plt
import numpy as np
import sklearn as sk
import tqdm
from nltk.stem.porter import PorterStemmer
from nltk.util import ngrams
from scipy.sparse import csr_matrix

## Loading the data

**Download the sentiment lexicon and the movie reviews dataset.**

**Load the movie reviews.**

Each word in a review comes with its part-of-speech tag. For documentation on POS-tags, see https://catalog.ldc.upenn.edu/docs/LDC99T42/tagguid1.pdf.


In [21]:
# file structure:
# [
#  {"cv": integer, "sentiment": str, "content": list}
#  {"cv": integer, "sentiment": str, "content": list}
#   ..
# ]
# where `content` is a list of sentences,
# with a sentence being a list of (token, pos_tag) pairs.


with open("reviews.json", mode="r", encoding="utf-8") as f:
    reviews = json.load(f)

print("Total number of reviews:", len(reviews), "\n")


def print_sentence_with_pos(s):
    print(" ".join("%s/%s" % (token, pos_tag) for token, pos_tag in s))


for i, r in enumerate(reviews):
    print(r["cv"], r["sentiment"], len(r["content"]))  # cv, sentiment, num sents
    print_sentence_with_pos(r["content"][0])
    if i == 4:
        break

c = Counter()
for review in reviews:
    for sentence in review["content"]:
        for token, pos_tag in sentence:
            c[token.lower()] += 1

print("\nNumber of word types:", len(c))
print("Number of word tokens:", sum(c.values()))

print("\nMost common tokens:")
for token, count in c.most_common(20):
    print("%10s : %8d" % (token, count))

Total number of reviews: 2000 

0 NEG 29
Two/CD teen/JJ couples/NNS go/VBP to/TO a/DT church/NN party/NN ,/, drink/NN and/CC then/RB drive/NN ./.
1 NEG 11
Damn/JJ that/IN Y2K/CD bug/NN ./.
2 NEG 24
It/PRP is/VBZ movies/NNS like/IN these/DT that/WDT make/VBP a/DT jaded/JJ movie/NN viewer/NN thankful/JJ for/IN the/DT invention/NN of/IN the/DT Timex/NNP IndiGlo/NNP watch/NN ./.
3 NEG 19
QUEST/NN FOR/IN CAMELOT/NNP ``/`` Quest/NNP for/IN Camelot/NNP ''/'' is/VBZ Warner/NNP Bros./NNP '/POS first/JJ feature-length/JJ ,/, fully-animated/JJ attempt/NN to/TO steal/VB clout/NN from/IN Disney/NNP 's/POS cartoon/NN empire/NN ,/, but/CC the/DT mouse/NN has/VBZ no/DT reason/NN to/TO be/VB worried/VBN ./.
4 NEG 38
Synopsis/NNPS :/: A/DT mentally/RB unstable/JJ man/NN undergoing/VBG psychotherapy/NN saves/VBZ a/DT boy/NN from/IN a/DT potentially/RB fatal/JJ accident/NN and/CC then/RB falls/VBZ in/IN love/NN with/IN the/DT boy/NN 's/POS mother/NN ,/, a/DT fledgling/NN restauranteur/NN ./.

Number of wo

# Lexicon-based approach (4pts)



In [99]:
from pathlib import Path

cwd_files = [f.name for f in Path.cwd().iterdir() if f.is_file()]

if "sent_lexicon" not in cwd_files:
    # download sentiment lexicon
    !wget https://gist.githubusercontent.com/bastings/d6f99dcb6c82231b94b013031356ba05/raw/f80a0281eba8621b122012c89c8b5e2200b39fd6/sent_lexicon
if "reviews.json" not in cwd_files:
    # download review data
    !wget https://gist.githubusercontent.com/bastings/d47423301cca214e3930061a5a75e177/raw/5113687382919e22b1f09ce71a8fecd1687a5760/reviews.json


A traditional approach to automatically classify documents according to their sentiment is the lexicon-based approach. To implement this approach, you need a **sentiment lexicon**, i.e., a list of words annotated with a sentiment label (e.g., positive and negative) or a sentiment score (e.g., a score from 0 to 5).

In this practical, you will use the sentiment
lexicon released by Wilson et al. (2005). The path of the loaded lexicon is `"sent_lexicon"`.

> Theresa Wilson, Janyce Wiebe, and Paul Hoffmann
(2005). [Recognizing Contextual Polarity in Phrase-Level Sentiment
Analysis](http://www.aclweb.org/anthology/H/H05/H05-1044.pdf). HLT-EMNLP.

Pay attention to all the information available in the sentiment lexicon. The field *word1* contains the lemma, *priorpolarity* contains the sentiment label (positive, negative, both, or neutral), *type* gives you the magnitude of the word's sentiment (strong or weak), and *pos1* gives you the part-of-speech tag of the lemma. Some lemmas can have multiple part-of-speech tags and thus multiple entries in the lexicon. For those lemmas you are free to use any entry, so it is fine to only keep one, e.g. the first or last one.

In [22]:
with open("sent_lexicon", mode="r", encoding="utf-8") as f:
    line_cnt = 0
    for line in f:
        print(line.strip())
        line_cnt += 1
        if line_cnt > 4:
            break

type=weaksubj len=1 word1=abandoned pos1=adj stemmed1=n priorpolarity=negative
type=weaksubj len=1 word1=abandonment pos1=noun stemmed1=n priorpolarity=negative
type=weaksubj len=1 word1=abandon pos1=verb stemmed1=y priorpolarity=negative
type=strongsubj len=1 word1=abase pos1=verb stemmed1=y priorpolarity=negative
type=strongsubj len=1 word1=abasement pos1=anypos stemmed1=y priorpolarity=negative


Given such a sentiment lexicon, there are ways to solve
the classification task without using Machine Learning. For example, one might look up every word $w_1 ... w_n$ in a document, and compute a **binary score**
$S_{binary}$ by counting how many words have a positive or a
negative label in the sentiment lexicon $SLex$.

$$S_{binary}(w_1 w_2 ... w_n) = \sum_{i = 1}^{n}\text{sign}(SLex\big[w_i\big])$$

where $\text{sign}(SLex\big[w_i\big])$ refers to the polarity of $w_i$.

**Threshold.** On average, there are more positive than negative words per review (~7.13 more positive than negative per review) to take this bias into account you should use a threshold of **8** (roughly the bias itself) to make it harder to classify as positive.

$$
\text{classify}(S_{binary}(w_1 w_2 ... w_n)) = \bigg\{\begin{array}{ll}
        \text{positive} & \text{if } S_{binary}(w_1w_2...w_n) > threshold\\
        \text{negative} & \text{otherwise}
        \end{array}
$$


#### (Q1.1) Implement this classifier using the provided structure below. (1 pt)

In [23]:
class SentimentLexicon:
    def __init__(
        self,
        lexicon_path: str = "./sent_lexicon",
        threshold: float = 8.0,
        weight: float = 1.0,
    ) -> None:
        """Sentiment classifier using a lexicon approach.

        Args:
            lexicon_path (str, optional): the location of the sentiment lexicon file. Defaults to "./sent_lexicon".
            threshold (float, optional): the threshold used in classification. Defaults to 8.0.
            weight (float, optional): the weight assigned to words with strong polarity. Defaults to 1.0.
        """
        self.lexicon_path = lexicon_path
        self.threshold = threshold
        self.weight = weight

        self.lexicon: dict[str, dict[str, str]]

        ##################
        self.lexicon = {}
        with open(self.lexicon_path, mode="r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()  # we split the text into individual lines
                entries = {}
                parts = line.split()
                for entry in parts:
                    if "=" in entry:
                        key, value = entry.split("=", 1)
                        entries[key] = value

                word = entries.get("word1")  # we retrieve the individual dimensions
                polarity = entries.get("priorpolarity")
                strength = entries.get("type")

                if not word or polarity in {
                    "neutral",
                    "both",
                }:  # we dont store neutral or ambiguous words in the lexicon
                    continue

                # we encode score based on polarity
                if polarity == "positive":
                    score = 1
                else:  # for negative words
                    score = -1
                if strength == "strongsubj":
                    score *= self.weight  # we adjust the weight
                # we store the words
                self.lexicon[word.lower()] = score  # we convert to words to their lower case format
        ##################

    def fit(self, documents: list[list[str]], labels: list[typing.Literal["POS", "NEG"]]) -> None:
        """Fit the classifer according to the input features and target labels.

        For the `SentimentLexicon` classifier, there are no parameters to learn.

        Args:
            documents (list[list[str]]): the list of documents as a list of tokens.
            labels (list[str]): the list of labels
        """
        pass

    def predict(self, documents: list[list[str]]) -> list[typing.Literal["POS", "NEG"]]:
        """Perform classification on input documents.

        Args:
            documents (list[list[str]]): the list of documents as a list of tokens.

        Returns:
            list[str]: predicted labels.
        """
        ##################
        predictions = []
        for doc in tqdm.tqdm(documents):
            total_score = 0
            for word in doc:
                word = word.lower()  # we convert to lower case
                if word in self.lexicon:
                    total_score += self.lexicon[word]
                else:
                    continue
            if total_score > self.threshold:
                predictions.append("POS")
            else:
                predictions.append("NEG")
        return predictions
        ##################

#### (Q1.2) Implement functions to transform the dataset to the needed data structures. (0.5pt)

In our current implementation, the classifier's `fit` and `predict` methods expect a list of corpus documents represented as `list[list[str]]`, and labels as a `list[str]`. However, our data is represented as `list[dict[str, Any]]`. Before we can fit and classify, we'll need to define a function to transform our data to the desired data structure.

In [24]:
def extract_labels(
    documents: list[dict[str, typing.Any]],
) -> list[typing.Literal["POS", "NEG"]]:
    """Converts a list of reviews to a list of labels.

    Args:
        documents (list[dict[str, Any]]): the reviews as a list of dicts.

    Returns:
        list[str]: the labels as a list of str
    """
    ##################
    labels = []
    for doc in documents:
        labels.append(doc["sentiment"])
    return labels
    ##################


def extract_unigrams(documents: list[dict[str, typing.Any]], lower: bool = True) -> list[list[str]]:
    """Converts a list of reviews to a list of unigram token lists.

    Args:
        documents (list[dict[str, typing.Any]]): the reviews as a list of dicts.
        lower (bool, optional): whether to lowercase the review tokens. Defaults to True.

    Returns:
        list[list[str]]: the list of unigram tokens
    """
    ##################
    all_unigrams = []  # we initialize a list collecting asll unigrams in all documents
    for doc in documents:
        unigrams_per_doc = (
            []
        )  # we intialize a list for the unigrams appearing in the specific document
        for sentence in doc["content"]:  # we loop over the sentences in the individual docs
            for word, pos in sentence:
                if lower == True:  # if we want to lowercase the review tokens
                    word = word.lower()
                unigrams_per_doc.append(word)
        all_unigrams.append(unigrams_per_doc)
    return all_unigrams
    ##################

#### (Q1.3) Implement a function to compute the accuracy of a classification. (0.5pt)

Finally, we need to implement a classification evaluation metric. For this practical, we'll be using [accuracy](https://en.wikipedia.org/wiki/Accuracy_and_precision#In_classification), which is defined as the proportion of correct classifications.

In [25]:
def accuracy(y_pred: list[str], y_true: list[str]) -> float:
    """Computes the accuracy score.

    Args:
        y_pred (list[str]): the predicted labels
        y_true (list[str]): the ground truth labels

    Returns:
        float: the accuracy score
    """
    ##################
    true_positives = 0
    for i in range(len(y_pred)):
        if y_pred[i] == y_true[i]:
            true_positives += 1
    accuracy = true_positives / len(y_pred)
    return accuracy
    ##################

Now train, predict and evaluate your classifier using the implemented functions. Make sure to print the final accuracy score.

In [26]:
##################
# First, we extract the labels and tokens
labels = extract_labels(reviews)
tokens = extract_unigrams(reviews)

# Second, we initialize our classifier
classifier = SentimentLexicon(
    lexicon_path="./sent_lexicon", threshold=8.0, weight=1.0
)  # with the predefined threshold, and basline weight

# For third step, we could fit the classifier, but since we dont have learnable parameters, we skip this step
# Now, we predict sentiment for the documents
predictions = classifier.predict(tokens)

# At last, we derive accuracy
performance = accuracy(predictions, labels)

print(f"The accuracy of the Lexicon-based classifier on our dataset is {performance}.")
##################

100%|██████████| 2000/2000 [00:00<00:00, 14517.56it/s]

The accuracy of the Lexicon-based classifier on our dataset is 0.6785.


#### (Q1.4) Now incorporate magnitude information and report the classification accuracy. Don't forget to use the threshold. (1 pt)

As the sentiment lexicon also has information about the **magnitude** of
sentiment (e.g., *“excellent"* would have higher magnitude than
*“good"*), we can take a more fine-grained approach by adding up all
sentiment scores, and deciding the polarity of the movie review using
the sign of the weighted score $S_{weighted}$.

$$S_{weighted}(w_1w_2...w_n) = \sum_{i = 1}^{n}SLex\big[w_i\big]$$

In [27]:
# Second, we initialize our classifier with a weight for magnitude
classifier_weighted = SentimentLexicon(
    lexicon_path="./sent_lexicon", threshold=8.0, weight=2.0
)  # with the predefined threshold, and adjusted higher weight

# For third step, we could fit the classifier, but since we don't have learnable parameters, we skip this step
# Now, we predict sentiment for the documents
predictions = classifier_weighted.predict(tokens)

# At last, we derive accuracy
performance_weighted = accuracy(predictions, labels)

print(
    "The accuracy of the weighted Lexicon-based classifier on our dataset is"
    f" {performance_weighted}."
)
##################

100%|██████████| 2000/2000 [00:00<00:00, 15281.24it/s]

The accuracy of the weighted Lexicon-based classifier on our dataset is 0.6865.


#### (Q1.5) A better threshold (1pt)

Above we have defined a threshold to account for an inherent bias in the dataset: there are more positive than negative words per review.
However, that threshold does not take into account *document length*. Explain why this is a problem and implement an alternative way to compute the threshold.

\####################
The underlying assumption is, that the document lengths are fairly similar. The observed bias is, that the proportion of negative/postive words is fairly constant, with a higher bias towards positive words. When the document length is uniform, and hence the number of positve word occurences is relatively uniform too, then this bias can be expressed as an average of 8 more positive word occurances per document. However, with the document length varying (and hence the number of positive words varying), the proportion between the negative/postive words is constant, not their difference. While for documents with 16 positive words we expect 8 negative word (the proportion is 0.5), for a document with 32 positive words we would not expect 24 negative words, but rather 16 (the proportion of 0.5 should be constant, not the difference of 8).
Consequently, we need to adopt a threshold, which takes into account the varying document length/number of sentiment word occurances, to succesfully model the constant proportion instead of the difference)

\####################

Now implement your agorithm by adjusting the previously created `SentimentLexicon` class. Make sure to print the final accuracy score and compare to the previous results.

Note: you might need to adjust your threshold.

In [28]:
class BetterSentimentLexicon(SentimentLexicon):
    def predict(self, documents: list[list[str]]) -> list[str]:
        """Perform classification on input documents.

        Args:
            documents (list[list[str]]): the list of documents as a list of tokens.

        Returns:
            list[str]: predicted labels.
        """
        ##################
        predictions = []
        for doc in tqdm.tqdm(documents):
            total_score = 0
            for word in doc:
                word = word.lower()  # we convert to lower case
                if word in self.lexicon:
                    total_score += self.lexicon[word]
                else:
                    continue
            # we normalize the total score, to incorporate document length
            normalized_score = total_score / len(doc)
            if (
                normalized_score > self.threshold
            ):  # in this case, threshold should be a float, incorporating the proportion in bias
                predictions.append("POS")
            else:
                predictions.append("NEG")
        return predictions
        ##################

In [29]:
##################

# Second, we initialize our classifier with a weight for magnitude
classifier_thresh = BetterSentimentLexicon(
    lexicon_path="./sent_lexicon", threshold=0.02, weight=2.0
)  # with the adjusted threshold depicting bias in terms of proportion, and adjusted higher weight

# For third step, we could fit the classifier, but since we don't have learnable parameters, we skip this step
# Now, we predict sentiment for the documents
predictions = classifier_thresh.predict(tokens)

# At last, we derive accuracy
performance_thresh = accuracy(predictions, labels)

print(
    f"The accuracy of the weighted Lexicon-based classifier on our dataset is {performance_thresh}."
)
##################

##################

100%|██████████| 2000/2000 [00:00<00:00, 14793.60it/s]

The accuracy of the weighted Lexicon-based classifier on our dataset is 0.69.


# Naive Bayes (10pt)

Your second task is to program a simple Machine Learning approach that operates
on a simple Bag-of-Words (BoW) representation of the text data, as
described by Pang et al. (2002). In this approach, the only features we
will consider are the words in the text themselves, without bringing in
external sources of information. The BoW model is a popular way of
representing texts as vectors, making it
easy to apply classical Machine Learning algorithms on NLP tasks.
However, the BoW representation is also very crude, since it discards
all information related to word order and grammatical structure in the
original text—as the name suggests.

## Writing your own classifier (4pts)

Write your own code to implement the Naive Bayes (NB) classifier. As
a reminder, the Naive Bayes classifier works according to the following
equation:
$$\hat{c} = \operatorname*{arg\,max}_{c \in C} P(c|\bar{f}) = \operatorname*{arg\,max}_{c \in C} P(c)\prod^n_{i=1} P(f_i|c)$$
where $C = \{ \text{POS}, \text{NEG} \}$ is the set of possible classes,
$\hat{c} \in C$ is the most probable class, and $\bar{f}$ is the feature
vector. Remember that we use the log of these probabilities when making
a prediction:
$$\hat{c} = \operatorname*{arg\,max}_{c \in C} \Big\{\log P(c) + \sum^n_{i=1} \log P(f_i|c)\Big\}$$

You can find more details about Naive Bayes in [Jurafsky &
Martin](https://web.stanford.edu/~jurafsky/slp3/). You can also look at
this helpful
[pseudo-code](https://nlp.stanford.edu/IR-book/html/htmledition/naive-bayes-text-classification-1.html).

*Note: this section and the next aim to put you in a position to replicate
    Pang et al.'s Naive Bayes results. However, your numerical results
    will differ from theirs, as they used different data.*

**You must write the Naive Bayes training and prediction code from
scratch.** You will not be given credit for using off-the-shelf Machine
Learning libraries.

The data contains the text of the reviews, where each document consists
of the sentences in the review, the sentiment of the review and an index
(cv) that you will later use for cross-validation. The
text has already been tokenised and POS-tagged for you. Your algorithm
should read in the text, **lowercase it**, store the words and their
frequencies in an appropriate data structure that allows for easy
computation of the probabilities used in the Naive Bayes algorithm, and
then make predictions for new instances.


#### (Q2.1) Unseen words (1pt)
The presence of words in the test dataset that
have not been seen during training can cause probabilities in the Naive Bayes classifier to equal $0$.
These can be words which are unseen in both positive and negative training reviews (case 1), but also words which are seen in reviews _of only one sentiment class_ in the training dataset (case 2). In both cases, **you should skip these words for both classes at test time**.  What would be the problem instead with skipping words only for one class in case 2?

\####################

The problem with skipping words for only one class in case 2 is that we would introduce a significant negative bias in our classifier for one class. First, by skipping the term "A" for Class 1 (for which the word "A" did not appear in the training data), the equation would miss the term log(P(A| Class 1)) for c=Class 1. However, without skipping it for Class 2, the equation would include the term log(P(A| Class 2)) for c=Class 2. Since 0<P(A|Class 2)=<1, log(P(A|CLass 2)=<0, when calculating


$$\Big\{\log P(Class 2) + \sum^n_{i=1} \log P(f_i|Class 2)\Big\}$$,

the equation would have an additional negative term for log(P(A|Class 2)), introducing a negative bias against Class 2. This bias is very counterintuitive, since the model is penalizing the class which had the evidence for a word occurence in contrast to the other class, that did not have that word occurance at all. Hence, to mitigate this negative and counterintuitive bias, the best approach is to skip this word A for both classes.

\####################

#### (Q2.2) Train your classifier on (positive and negative) reviews with cv-value 000-899, and test it on the remaining (positive and negative) reviews cv900–cv999.  Report results using classification accuracy as your evaluation metric. (2pts)


In [30]:
class NaiveBayes:
    def __init__(self, smoothing: float = 0):
        """Naive Bayes classifier for multinomial data.

        Args:
            smoothing (float, optional): the Laplacian smoothing factor. Defaults to 0.
        """
        self.smoothing = smoothing

        # The datastructures necessary for NaiveBayes
        self.vocab: set[str] = set()

        self.prior: dict[typing.Literal["POS", "NEG"], float] = {
            "POS": 0,
            "NEG": 0,
        }

        self.token_counts: dict[str, dict[typing.Literal["POS", "NEG"], int]] = defaultdict(
            lambda: {"POS": 0, "NEG": 0},
        )

        self.cond_prob: dict[str, dict[typing.Literal["POS", "NEG"], float]] = defaultdict(
            lambda: {"POS": 0.0, "NEG": 0.0},
        )

    def fit(
        self,
        documents: list[list[typing.Any]],
        labels: list[typing.Literal["POS", "NEG"]],
    ) -> None:
        """Fit the classifer according to the input features and target labels.

        Args:
            documents (list[list[str]]): the list of documents as a list of tokens
            labels (list[str]): the list of labels
        """
        ##################
        # First, we start by estimating the priors
        nr_documents = len(documents)
        nr_pos_doc = sum(1 for label in labels if label == "POS")
        nr_neg_doc = nr_documents - nr_pos_doc

        self.prior["POS"] = math.log(
            nr_pos_doc / nr_documents
        )  # the log prior for the positive and negative classes
        self.prior["NEG"] = math.log(nr_neg_doc / nr_documents)

        # Second, we estimate the conditional probabilities
        for words, label in zip(documents, labels):  # we start by creating the vocabulary
            for word in words:
                # Handle both strings and tuples (n-grams) for exercise 2.8
                if isinstance(word, str):
                    word = word.lower()
                # For tuples (n-grams), they are already lowercased in extract_ngrams

                self.vocab.add(word)
                self.token_counts[word][label] += 1

        # to derive the class conditional probailities, we calculate the total word counts for each class
        self.total_word_class = {"POS": 0, "NEG": 0}
        for word in self.vocab:
            self.total_word_class["POS"] += self.token_counts[word]["POS"]
            self.total_word_class["NEG"] += self.token_counts[word]["NEG"]
        # we compute the conditional probabilites with smoothing
        for word in self.vocab:
            for clss in ["POS", "NEG"]:
                numerator = self.token_counts[word][clss] + self.smoothing
                denominator = self.total_word_class[clss] + self.smoothing * len(self.vocab)
                self.cond_prob[word][clss] = (
                    numerator / denominator
                )  # conditional probabilities per word

        ##################

    def predict(self, documents: list[list[typing.Any]]) -> list[typing.Literal["POS", "NEG"]]:
        """Perform classification on input documents.

        Args:
            documents (list[list[str]]): the list of documents as a list of tokens

        Returns:
            list[str]: predicted labels
        """
        ##################
        # we initialize the predicted labels as an empty list
        pred_labels = []
        # We start by iterating over every document
        for document in documents:
            c_hat_pos = self.prior["POS"]
            c_hat_neg = self.prior["NEG"]
            for word in document:
                # Handle both strings and tuples (n-grams) for exercise 2.8
                if isinstance(word, str):
                    word = word.lower()
                # For tuples (n-grams), they are already lowercased in extract_ngrams

                if word not in self.vocab:  # we skip unseen words
                    continue
                if (
                    self.token_counts[word]["POS"] == 0 or self.token_counts[word]["NEG"] == 0
                ):  # we skip words only represented in one class
                    continue
                c_hat_pos += math.log(self.cond_prob[word]["POS"])
                c_hat_neg += math.log(self.cond_prob[word]["NEG"])
            if c_hat_pos > c_hat_neg:
                pred_labels.append("POS")
            else:
                pred_labels.append("NEG")
        return pred_labels
        ##################

Now train, predict and evaluate your classifier on the reviews dataset. Make sure to select the right reviews for the train and test sets.

In [31]:
##################

# we start by splitting our training data and test data
train_docs, train_labels = [], []
test_docs, test_labels = [], []
for doc, label in zip(reviews, labels):
    if doc["cv"] >= 0 and doc["cv"] <= 899:
        train_docs.append(doc)
        train_labels.append(label)
    elif doc["cv"] >= 900 and doc["cv"] <= 999:
        test_docs.append(doc)
        test_labels.append(label)

# now we extract the training and test tokens from the review dataset with the method introudced in part 1
train_tokens = extract_unigrams(train_docs)
test_tokens = extract_unigrams(test_docs)

# we initialize the classifier without smoothing
nb_classifier = NaiveBayes(smoothing=0)

# we fit it
nb_classifier.fit(train_tokens, train_labels)

# we derive the predictions
predictions = nb_classifier.predict(test_tokens)

# at last, we derive the accuracy with our function used previously
performance = accuracy(predictions, test_labels)
print(
    "The performance of the Naive Bayes classifier on our dataset with Laplacian smoothing of"
    f" {nb_classifier.smoothing} is {performance}."
)
##################

The performance of the Naive Bayes classifier on our dataset with Laplacian smoothing of 0 is 0.825.


#### (Q2.3) Would you consider accuracy to also be a good way to evaluate your classifier in a situation where 90% of your data instances are of positive movie reviews? (1pt)

Simulate this scenario by keeping the positive reviews
data unchanged, but only using negative reviews cv000–cv089 for
training, and cv900–cv909 for testing. Calculate the classification
accuracy, and explain what changed.

\####################

The accuracy metric is inherently a bad choice for evaluating classifiers on a skewed dataset with minority class. In our case, this minority class is Negative reviews. The theoretical argument against using accuracy as a describing metric in skewed datasets is that a trivial classifier that always predicts the majority class (in this case POS), would get an accuracy of the proportion of the majority class. In this case, a classifier predicting POS all the time would have a 90% accuracy evaluated on the skewed dataset. However, this model would not perform well on a novel dataset.

However, when I ran the experiment, the accuracy of the classifier dropped to 60 % from the intial 82.5%. this can be explained by the training process, and by taking a closer look at the confusion matrix of the predictions. We can infer, that the classifier did not predict only one class,but for the negative class it predicted negative class 90% of the time, and for positive class the accuracy was much lower around 57%. These results imply that the model overfit on the minority class, and failed to generalize to the majority class "POS". Hence, accuracy is not a good metric to describe model performance for skewed datasets, but for meaningful evaluations we need to include specific metrics, such as Precision, Recall as well.
\####################

In [32]:
##################
# we start by splitting our training data and test data
train_docs, train_labels = [], []
test_docs, test_labels = [], []
for doc, label in zip(reviews, labels):
    if label == "POS":
        if doc["cv"] >= 0 and doc["cv"] <= 899:
            train_docs.append(doc)
            train_labels.append(label)
        elif doc["cv"] >= 900 and doc["cv"] <= 999:
            test_docs.append(doc)
            test_labels.append(label)
    elif label == "NEG":
        if doc["cv"] >= 0 and doc["cv"] <= 89:
            train_docs.append(doc)
            train_labels.append(label)
        elif doc["cv"] >= 900 and doc["cv"] <= 909:
            test_docs.append(doc)
            test_labels.append(label)
# now we extract the training and test tokens from the review dataset with the method introudced in part 1
train_tokens = extract_unigrams(train_docs)
test_tokens = extract_unigrams(test_docs)

# we initialize the classifier without smoothing
nb_classifier = NaiveBayes(smoothing=0)

# we fit it
nb_classifier.fit(train_tokens, train_labels)

# we derive the predictions
predictions = nb_classifier.predict(test_tokens)

# at last, we derive the accuracy with our function used previously
performance = accuracy(predictions, test_labels)
print(
    "The performance of the Naive Bayes classifier on our skewed dataset with Laplacian smoothing"
    f" of {nb_classifier.smoothing} is {performance}."
)

##################

The performance of the Naive Bayes classifier on our skewed dataset with Laplacian smoothing of 0 is 0.6.


In [43]:
# for deeper analysis, we look at the confusion matrix as well
def confusion_matrix(y_true, y_pred, labels=("POS", "NEG")):
    cm = {a: {b: 0 for b in labels} for a in labels}
    for t, p in zip(y_true, y_pred):
        cm[t][p] += 1
    return cm

cm = confusion_matrix(test_labels, predictions)

# We print it in readable format
print("                    PREDICTED")
print("              POS        NEG")
print("GROUND TRUTH --------------------")
for true_label in ["POS", "NEG"]:
    row = f"{true_label:11s}"
    for pred_label in ["POS", "NEG"]:
        row += f"{cm[true_label][pred_label]:10d}"
    print(row)


                    PREDICTED
              POS        NEG
GROUND TRUTH --------------------
POS                57        43
NEG                 1         9


## Smoothing (1pt)

As mentioned above, the presence of words in the test dataset that
have not been seen during training can cause probabilities in the Naive
Bayes classifier to be $0$, thus making that particular test instance
undecidable. The standard way to mitigate this effect (as well as to
give more clout to rare words) is to use smoothing, in which the
probability fraction

$$\frac{\text{count}(w_i, c)}{\sum\limits_{w\in V} \text{count}(w, c)}$$

for a word $w_i$ becomes

$$\frac{\text{count}(w_i, c) + \text{smoothing}(w_i)}{\sum\limits_{w\in V} \text{count}(w, c) + \sum\limits_{w \in V} \text{smoothing}(w)}$$


#### (Q2.4) Implement Laplace feature smoothing (1pt)
Implement Laplace smoothing, i.e., smoothing with a constant value ($smoothing(w) = \kappa, \forall w \in V$), in your Naive
Bayes classifier’s code, and report the accuracy for $\kappa = 1$ and $\kappa = 3.5$. Train your classifier on (positive and negative) reviews with cv-value 000-899, and test it on the remaining (positive and negative) reviews cv900–cv999.

In [ ]:
##################

# we start by splitting our training data and test data
train_docs, train_labels = [], []
test_docs, test_labels = [], []
for doc, label in zip(reviews, labels):
    if doc["cv"] >= 0 and doc["cv"] <= 899:
        train_docs.append(doc)
        train_labels.append(label)
    elif doc["cv"] >= 900 and doc["cv"] <= 999:
        test_docs.append(doc)
        test_labels.append(label)

# now we extract the training and test tokens from the review dataset with the method introudced in part 1
train_tokens = extract_unigrams(train_docs)
test_tokens = extract_unigrams(test_docs)

# we initialize the classifier without smoothing
nb_classifier1 = NaiveBayes(smoothing=1)  # k=1
nb_classifier2 = NaiveBayes(smoothing=3.5)  # k=3.5

# we fit it
nb_classifier1.fit(train_tokens, train_labels)
nb_classifier2.fit(train_tokens, train_labels)

# we derive the predictions
predictions1 = nb_classifier1.predict(test_tokens)
predictions2 = nb_classifier2.predict(test_tokens)

# at last, we derive the accuracy with our function used previously
performance1 = accuracy(predictions1, test_labels)
performance2 = accuracy(predictions2, test_labels)

print(
    "The performance of the Naive Bayes classifier on our dataset with Laplacian smoothing of"
    f" {nb_classifier1.smoothing} is {performance1}."
)
print(
    "\n\nThe performance of the Naive Bayes classifier on our dataset with Laplacian smoothing of"
    f" {nb_classifier2.smoothing} is {performance2}."
)


##################

The performance of the Naive Bayes classifier on our dataset with Laplacian smoothing of 1 is 0.835.


The performance of the Naive Bayes classifier on our dataset with Laplacian smoothing of 3.5 is 0.835.


**FROM NOW ON, ALWAYS USE SMOOTHING (YOU CAN CHOOSE $\kappa$) WHEN USING THE NAIVE BAYES CLASSIFIER**

## Cross-Validation (1pt)

A serious danger in using Machine Learning on small datasets, with many
iterations of slightly different versions of the algorithms, is ending up with Type III errors, also called the “testing hypotheses
suggested by the data” errors. This type of error occurs when we make
repeated improvements to our classifiers by playing with features and
their processing, but we don’t get a fresh, never-before seen test
dataset every time. Thus, we risk developing a classifier that gets better
and better on our data, but only gets worse at generalizing to new, unseen data. In other words, we risk developping a classifier that overfits.

A simple method to guard against Type III errors is to use
Cross-Validation. In **N-fold Cross-Validation**, we divide the data into N
distinct chunks, or folds. Then, we repeat the experiment N times: each
time holding out one of the folds for testing, training our classifier
on the remaining N - 1 data folds, and reporting performance on the
held-out fold. We can use different strategies for dividing the data:

-   Consecutive splitting:
  - cv000–cv099 = Split 1
  - cv100–cv199 = Split 2
  - etc.
  
-   Round-robin splitting (mod 10):
  - cv000, cv010, cv020, … = Split 1
  - cv001, cv011, cv021, … = Split 2
  - etc.

-   Random sampling/splitting
  - Not used here (but you may choose to split this way in a non-educational situation)



#### (Q2.5) Write the code to implement 10-fold cross-validation using round-robin splitting for your Naive Bayes classifier from Q2.4 and compute the 10 accuracies. Report the mean and standard deviation of the accuracies. (1pt)

In [ ]:
def cross_validation_splits(
    documents: list[dict[str, typing.Any]],
    num_folds: int = 10,
) -> list[tuple[list, list]]:
    """Splits the reviews into disjoint subsets of train/test splits.

    Args:
        documents (list[dict[str, typing.Any]]): the reviews as dicts.
        num_folds (int, optional): the number of cross-validation folds. Defaults to 10.

    Returns:
        list[tuple[list, list]]: a list of train/test folds
    """
    ##################
    # Initialize list to hold all folds (train, test) pairs
    folds = []

    # For each fold, we assign which documents are test and which are train
    for fold_idx in range(num_folds):
        train_fold = []
        test_fold = []

        # Iterate through all documents and assign based on round-robin splitting
        # Documents with cv % num_folds == fold_idx go to test, others to train
        for doc in documents:
            if doc["cv"] % num_folds == fold_idx:
                test_fold.append(doc)
            else:
                train_fold.append(doc)

        folds.append((train_fold, test_fold))

    return folds
    ##################
    # raise NotImplementedError

Now re-train the Naive Bayes classifier on each fold, and evaluate on that fold's held-out test set. Make sure to print the accuracy mean and standard deviation.

Note: you may use `tqdm` to add progress bars

In [ ]:
##################

# Get the cross-validation splits
cv_splits = cross_validation_splits(reviews, num_folds=10)

# Store accuracies for each fold
fold_accuracies = []

# Perform 10-fold cross-validation
for fold_idx, (train_fold, test_fold) in enumerate(tqdm.tqdm(cv_splits, desc="Cross-validation")):
    # Extract labels
    fold_train_labels = extract_labels(train_fold)
    fold_test_labels = extract_labels(test_fold)

    # Extract unigram tokens
    fold_train_tokens = extract_unigrams(train_fold)
    fold_test_tokens = extract_unigrams(test_fold)

    # Initialize and train Naive Bayes classifier with smoothing
    nb_classifier_cv = NaiveBayes(smoothing=1.0)
    nb_classifier_cv.fit(fold_train_tokens, train_labels)

    # Make predictions
    predictions_fold = nb_classifier_cv.predict(fold_test_tokens)

    # Calculate accuracy for this fold
    fold_accuracy = accuracy(predictions_fold, fold_test_labels)  # type: ignore
    fold_accuracies.append(fold_accuracy)


# Calculate mean and standard deviation
mean_accuracy_nb = np.mean(fold_accuracies)
std_accuracy_nb = np.std(fold_accuracies)

print(f"\n{'='*50}")
print(f"10-Fold Cross-Validation Results:")
print(f"Mean Accuracy: {mean_accuracy_nb:.4f}")
print(f"Standard Deviation: {std_accuracy_nb:.4f}")
print(f"{'='*50}")

##################
# raise NotImplementedError

Cross-validation: 100%|██████████| 10/10 [00:03<00:00,  3.33it/s]


10-Fold Cross-Validation Results:
Mean Accuracy: 0.8240
Standard Deviation: 0.0282


**FROM NOW ON, ALWAYS USE CROSS VALIDATION WHEN EVALUATING A CLASSIFIER**

## Features, overfitting, and the curse of dimensionality

In the Bag-of-Words model, ideally we would like each distinct word in
the text to be mapped to its own dimension in the output vector
representation. However, real world text is messy, and we need to decide
on what we consider to be a word. For example, is “`word`" different
from “`Word`", from “`word`”, or from “`words`"? Too strict a
definition, and the number of features explodes, while our algorithm
fails to learn anything generalisable. Too lax, and we risk destroying
our learning signal. In the following section, you will learn about
confronting the feature sparsity and the overfitting problems as they
occur in NLP classification tasks.

### Stemming (1.5pts)

To make your algorithm more robust, use stemming and
hash different inflections of a word to the same feature in the BoW
vector space.


#### (Q2.6): How does the performance of your classifier change when you use stemming on your training and test datasets? (1pt)

First we'll need to adjust our `extract_unigrams` function to extract stemmed unigrams. Use the provided structure below.

In [ ]:
def extract_stems(
    documents: list[dict[str, typing.Any]], stemmer: PorterStemmer, lower: bool = True
) -> list[list[str]]:
    """Extract stemmed unigrams from a corpus of documents.

    Args:
        documents (list[dict[str, Any]]): the list of documents
        stemmer (PorterStemmer): the stemmer to use when stemming documents
        lower (bool, optional): whether to lowercase tokens before stemming. Defaults to True.

    Returns:
        list[list[str]]: the list of token lists
    """
    ##################

    # Initialize list to collect all stemmed tokens from all documents
    all_stems = []

    for doc in documents:
        # Initialize list for stemmed tokens in this specific document
        stems_per_doc = []

        # Loop over sentences in the document
        for sentence in doc["content"]:

            # Extract word and POS tag from each token
            for word, pos in sentence:
                # Lowercase the word if requested
                if lower:
                    word = word.lower()

                # Stem the word using the provided stemmer
                stemmed_word = stemmer.stem(word)
                stems_per_doc.append(stemmed_word)

        all_stems.append(stems_per_doc)

    return all_stems

    ##################
    # raise NotImplementedError

Now re-train your Naive Bayes classifier. Use cross-validation to evaluate the classifier. Use nltk's `PorterStemmer` to stem the document tokens.


In [ ]:
##################
# Initialize the Porter Stemmer
stemmer = PorterStemmer()

# Store accuracies for each fold
fold_accuracies_stemmed = []

# Perform 10-fold cross-validation with stemming
for fold_idx, (train_fold, test_fold) in enumerate(
    tqdm.tqdm(cv_splits, desc="Cross-validation with stemming")
):
    # Extract labels
    fold_train_labels = extract_labels(train_fold)
    fold_test_labels = extract_labels(test_fold)

    # Extract STEMMED tokens (this is the key difference from Q2.5)
    fold_train_stems = extract_stems(train_fold, stemmer=stemmer, lower=True)
    fold_test_stems = extract_stems(test_fold, stemmer=stemmer, lower=True)

    # Initialize and train Naive Bayes classifier with smoothing
    nb_classifier_stem = NaiveBayes(smoothing=1.0)
    nb_classifier_stem.fit(fold_train_stems, fold_train_labels)

    # Make predictions
    predictions_fold = nb_classifier_stem.predict(fold_test_stems)

    # Calculate accuracy for this fold
    fold_accuracy = accuracy(predictions_fold, fold_test_labels)  # type: ignore
    fold_accuracies_stemmed.append(fold_accuracy)

# Calculate mean and standard deviation
mean_accuracy_stemmed = np.mean(fold_accuracies_stemmed)
std_accuracy_stemmed = np.std(fold_accuracies_stemmed)

print(f"\n{'='*50}")
print(f"10-Fold Cross-Validation Results (with Stemming):")
print(f"Mean Accuracy: {mean_accuracy_stemmed:.4f}")
print(f"Standard Deviation: {std_accuracy_stemmed:.4f}")
print(f"{'='*50}")

##################
# raise NotImplementedError

Cross-validation with stemming: 100%|██████████| 10/10 [00:51<00:00,  5.11s/it]


10-Fold Cross-Validation Results (with Stemming):
Mean Accuracy: 0.8185
Standard Deviation: 0.0292


#### (Q2.7) What happens to the number of features (i.e., the size of the vocabulary) when using stemming as opposed to (Q2.4)? (0.5pt)

Give actual numbers. You can use the training set from Q2.4 to determine these.

In [ ]:
##################

# Build vocabulary by collecting all unique tokens from the Q2.4 train tokens
vocab_no_stem = set()
for doc_tokens in train_tokens:
    vocab_no_stem.update(doc_tokens)

vocab_size_no_stem = len(vocab_no_stem)

# Get vocabulary size with stemming
stemmer = PorterStemmer()
train_tokens_with_stem = extract_stems(train_docs, stemmer=stemmer, lower=True)

# Build vocabulary by collecting all unique stems
vocab_with_stem = set()
for doc_stems in train_tokens_with_stem:
    vocab_with_stem.update(doc_stems)

vocab_size_with_stem = len(vocab_with_stem)

# Calculate the reduction
reduction = vocab_size_no_stem - vocab_size_with_stem
reduction_percent = (reduction / vocab_size_no_stem) * 100

print(f"Vocabulary size With stemming: {vocab_size_with_stem}")
print(f"Vocabulary size Without stemming: {vocab_size_no_stem}")
print(f"Reduction in features: {reduction} ({reduction_percent:.2f}%)")

##################
# raise NotImplementedError

Vocabulary size With stemming: 32404
Vocabulary size Without stemming: 45348
Reduction in features: 12944 (28.54%)


### N-grams (2.5pts)

A simple way of retaining some of the word
order information when using bag-of-words representations is to add **n-gram** features.






#### (Q2.8) Retrain your classifier from (Q2.4) using **unigrams+bigrams** and **unigrams+bigrams+trigrams** as features. (2pt)

We'll need to adjust our `extract_unigrams` function to extract stemmed unigrams. Use the provided structure below. You can use nltk's `ngrams` function to extract n-grams from the sentences.

In [ ]:
def extract_ngrams(
    documents: list[dict], n_values: typing.Sequence[int], lower: bool = True
) -> list[list[typing.Any]]:
    """Extract n-gram features from the corpus of documents.

    Args:
        documents (list[dict]): the list of documents
        n_values (typing.Sequence[int]): the different sizes of n-grams to take (1=unigram, 2=bigram)
        lower (bool, optional): whether to lowercase the tokens. Defaults to True.

    Returns:
        list[list[tuple[str]]]: the n-gram tokens
    """
    ##################

    # Initialize list to collect all n-grams from all documents
    all_ngrams = []

    # Iterate over each document
    for doc in documents:
        # Initialize list for n-grams in this specific document
        ngrams_per_doc = []

        # Loop over sentences in the document
        for sentence in doc["content"]:
            # Extract just the words (not POS tags) from the sentence
            words = [word for word, _ in sentence]

            # Lowercase if requested
            if lower:
                words = [word.lower() for word in words]

            # For each n-gram size requested (e.g., [1, 2] for unigrams and bigrams)
            for n in n_values:
                # Use nltk's ngrams function to extract n-grams
                sentence_ngrams = list(ngrams(words, n))

                # Add the n-grams from this sentence to the document's n-grams
                ngrams_per_doc.extend(sentence_ngrams)

        # Add this document's n-grams to the collection
        all_ngrams.append(ngrams_per_doc)

    return all_ngrams

    ##################
    # raise NotImplementedError

Now re-train your Naive Bayes classifier with **uni- and bigram** tokens. Report accuracy and compare it with that of the approaches you have previously implemented. Use cross-validation when evaluating your classifier. (1pt)

In [ ]:
##################

# Store accuracies for each fold
fold_accuracies_bigrams = []

# Perform 10-fold cross-validation with unigrams + bigrams
for fold_idx, (train_fold, test_fold) in enumerate(
    tqdm.tqdm(cv_splits, desc="Cross-validation with unigrams+bigrams")
):
    # Extract labels
    fold_train_labels = extract_labels(train_fold)
    fold_test_labels = extract_labels(test_fold)

    # Extract unigram + bigram features using n_values=[1, 2]
    fold_train_ngrams = extract_ngrams(train_fold, n_values=[1, 2], lower=True)
    fold_test_ngrams = extract_ngrams(test_fold, n_values=[1, 2], lower=True)

    # Initialize and train Naive Bayes classifier with smoothing
    nb_classifier_bigrams = NaiveBayes(smoothing=1.0)
    nb_classifier_bigrams.fit(fold_train_ngrams, fold_train_labels)

    # Make predictions
    predictions_fold = nb_classifier_bigrams.predict(fold_test_ngrams)

    # Calculate accuracy for this fold
    fold_accuracy = accuracy(predictions_fold, fold_test_labels)  # type: ignore
    fold_accuracies_bigrams.append(fold_accuracy)

# Calculate mean and standard deviation
mean_accuracy_bigrams = np.mean(fold_accuracies_bigrams)
std_accuracy_bigrams = np.std(fold_accuracies_bigrams)

print(f"\n{'='*50}")
print(f"10-Fold Cross-Validation Results (Unigrams + Bigrams):")
print(f"Mean Accuracy: {mean_accuracy_bigrams:.4f}")
print(f"Standard Deviation: {std_accuracy_bigrams:.4f}")
print(f"{'='*50}")


##################
# raise NotImplementedError

Cross-validation with unigrams+bigrams: 100%|██████████| 10/10 [00:19<00:00,  1.99s/it]


10-Fold Cross-Validation Results (Unigrams + Bigrams):
Mean Accuracy: 0.8290
Standard Deviation: 0.0286


Now re-train your Naive Bayes classifier with **uni-, bi- and trigram** tokens. Report accuracy and compare it with that of the approaches you have previously implemented. Use cross-validation when evaluating your classifier. (1pt)

In [ ]:
##################
# Store accuracies for each fold
fold_accuracies_trigrams = []

# Perform 10-fold cross-validation with unigrams + bigrams + trigrams
for fold_idx, (train_fold, test_fold) in enumerate(
    tqdm.tqdm(cv_splits, desc="Cross-validation with unigrams+bigrams+trigrams")
):
    # Extract labels
    fold_train_labels = extract_labels(train_fold)
    fold_test_labels = extract_labels(test_fold)

    # Extract unigram + bigram + trigram features using n_values=[1, 2, 3]
    fold_train_ngrams = extract_ngrams(train_fold, n_values=[1, 2, 3], lower=True)
    fold_test_ngrams = extract_ngrams(test_fold, n_values=[1, 2, 3], lower=True)

    # Initialize and train Naive Bayes classifier with smoothing
    nb_classifier_trigrams = NaiveBayes(smoothing=1.0)
    nb_classifier_trigrams.fit(fold_train_ngrams, fold_train_labels)

    # Make predictions
    predictions_fold = nb_classifier_trigrams.predict(fold_test_ngrams)

    # Calculate accuracy for this fold
    fold_accuracy = accuracy(predictions_fold, fold_test_labels)  # type: ignore
    fold_accuracies_trigrams.append(fold_accuracy)

# Calculate mean and standard deviation
mean_accuracy_trigrams = np.mean(fold_accuracies_trigrams)
std_accuracy_trigrams = np.std(fold_accuracies_trigrams)

print(f"\n{'='*50}")
print(f"10-Fold Cross-Validation Results (Unigrams + Bigrams + Trigrams):")
print(f"Mean Accuracy: {mean_accuracy_trigrams:.4f}")
print(f"Standard Deviation: {std_accuracy_trigrams:.4f}")
print(f"{'='*50}")

##################
# raise NotImplementedError

Cross-validation with unigrams+bigrams+trigrams: 100%|██████████| 10/10 [00:43<00:00,  4.31s/it]


10-Fold Cross-Validation Results (Unigrams + Bigrams + Trigrams):
Mean Accuracy: 0.8145
Standard Deviation: 0.0288


#### (Q2.9): How many features does the BoW model have to take into account now? (0.5pt)
How would you expect the number of features to increase theoretically (e.g., linear, square, cubed, exponential)? How do the number of features increase in the held-out training set (compared to Q2.8)? Do you expect this rate of increase to continue for (much) larger n-grams?

Use the held-out training set from training set from Q2.4 for this.


\####################

Theoretically we would expect exponentioal increase. For V vocabulary size and n n-gram size each position can have V possibilities. This leads to V^2 bigrams and V^3 trigrams.

In practice we see much slower growth due to data sparsity, because many theoretically possible n-gram combinations don't occur in real text. We expect this growth rate to further decrease drastically in larger n-grams sizes for the same reasons.

\####################

In [ ]:
##################

# Extract different n-gram features from the training set
# TODO Question TA on did they mean the training set from Q2.4 as used in previous questions or the test set here (bc. of held-out phrasing)
train_unigrams = extract_ngrams(train_docs, n_values=[1], lower=True)
train_bigrams = extract_ngrams(train_docs, n_values=[1, 2], lower=True)
train_trigrams = extract_ngrams(train_docs, n_values=[1, 2, 3], lower=True)

# Build vocabularies
vocab_unigrams = set()
for doc in train_unigrams:
    vocab_unigrams.update(doc)

vocab_bigrams = set()
for doc in train_bigrams:
    vocab_bigrams.update(doc)

vocab_trigrams = set()
for doc in train_trigrams:
    vocab_trigrams.update(doc)

# Calculate sizes and growth
size_unigrams = len(vocab_unigrams)
size_bigrams = len(vocab_bigrams)
size_trigrams = len(vocab_trigrams)

increase_bigrams = size_bigrams - size_unigrams
increase_trigrams = size_trigrams - size_bigrams

growth_rate_2gram = increase_bigrams / size_unigrams
growth_rate_3gram = increase_trigrams / size_bigrams

# Print results
print(f"\n{'='*50}")
print(f"Unigrams: {size_unigrams:,}")
print(f"Unigrams + Bigrams: {size_bigrams:,} (+{growth_rate_2gram:.2%})")
print(f"Uni + Bi + Trigrams: {size_trigrams:,} (+{growth_rate_3gram:.2%})")
print(f"{'='*50}")
##################
# raise NotImplementedError


Unigrams: 45,348
Unigrams + Bigrams: 465,262 (+925.98%)
Uni + Bi + Trigrams: 1,346,107 (+189.32%)





# (3) Support Vector Machines (4pts)

Though simple to understand, implement, and debug, one major problem with the Naive Bayes classifier is that its performance deteriorates (becomes skewed) when it is being used with features which are not independent (i.e., are correlated). Another popular classifier that doesn’t assume feature independence is the Support Vector Machine (SVM) classifier.

You can find more details about SVMs in Chapter 7 of Bishop: Pattern Recognition and Machine Learning.
Other sources for learning SVM:
* http://web.mit.edu/zoya/www/SVM.pdf
* http://www.cs.columbia.edu/~kathy/cs4701/documents/jason_svm_tutorial.pdf
* https://pythonprogramming.net/support-vector-machine-intro-machine-learning-tutorial/


#### (Q3.1): Train SVM and compare to Naive Bayes (2pts)

Train an SVM classifier (`sklearn.svm.LinearSVC`) using the unigram features collected for Naive Bayes.

sklearn's classifiers expect a different data structure than what we've been using so far. Instead of a list of tokens, we'll need to provide a (sparse) [document-term matrix](https://en.wikipedia.org/wiki/Document-term_matrix). Each row represents a document, and each column represents a token. The value in each cell represents how often a token appears in a document.

Define a function below that takes a list of tokens and constructs a document-term matrix. While not mandatory, it's recommended to use scipy's `csr_matrix` to produce a sparse matrix representation. This avoids having to keep many 0 values in memory, and can speed up SVM training.

Hint: the documentation on the [`csr_matrix`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html) is very helpful

In [ ]:
def build_term_document_matrix(
    train_features: list[list[str]], test_features: list[list[str]]
) -> tuple[np.array, np.array]:
    """Converts a list of token lists to a document-term matrix.

    Args:
        train_features (list[list[str]]): the training token lists
        test_features (list[list[str]]): the testing token lists

    Returns:
        tuple[array, array]: a tuple of training and testing DTMs
    """
    ##################

    # Build vocabulary from training data
    vocab = {}
    vocab_index = 0
    for doc in train_features:
        for token in doc:
            if token not in vocab:
                vocab[token] = vocab_index
                vocab_index += 1

    # Build training matrix
    train_rows = []
    train_cols = []
    train_data = []

    for doc_idx, doc in enumerate(train_features):
        token_counts = Counter(doc)
        for token, count in token_counts.items():
            if token in vocab:  # Should always be true for train
                train_rows.append(doc_idx)
                train_cols.append(vocab[token])
                train_data.append(count)

    train_matrix = csr_matrix(
        (train_data, (train_rows, train_cols)), shape=(len(train_features), len(vocab))
    )

    # Build test matrix using same vocabulary
    test_rows = []
    test_cols = []
    test_data = []

    for doc_idx, doc in enumerate(test_features):
        token_counts = Counter(doc)
        for token, count in token_counts.items():
            if token in vocab:  # Only use tokens seen in training
                test_rows.append(doc_idx)
                test_cols.append(vocab[token])
                test_data.append(count)

    test_matrix = csr_matrix(
        (test_data, (test_rows, test_cols)), shape=(len(test_features), len(vocab))
    )

    return (train_matrix, test_matrix)

    ##################
    # raise NotImplementedError

Besides a document term matrix, sklearn also expects the labels to be a list of `int`. Using the structure provide, implement a function to convert the `str` labels to integers.

In [ ]:
def convert_labels_to_ints(labels: list[str]) -> list[int]:
    """Converts a list of "POS" or "NEG" to 0 or 1.

    Args:
        labels (list[str]): the list of str labels

    Returns:
        list[int]: the list of int labels
    """
    ##################

    return [1 if label == "POS" else 0 for label in labels]

    ##################
    # raise NotImplementedError

Now train, predict and evaluate an SVM classifier. Compare the classification performance of the SVM classifier to that of the Naive Bayes classifier with smoothing. Use cross-validation to evaluate the performance of the classifiers.

You are not required to perform hyperparameter tuning, but it might be helpful.

In [ ]:
##################

# Store accuracies for both classifiers across folds
fold_accuracies_svm = []

# Perform 10-fold cross-validation
for fold_idx, (train_fold, test_fold) in enumerate(
    tqdm.tqdm(cv_splits, desc="Cross-validation: SVM vs NB")
):
    # Extract labels
    fold_train_labels = extract_labels(train_fold)
    fold_test_labels = extract_labels(test_fold)

    # Extract unigram tokens
    fold_train_tokens = extract_unigrams(train_fold, lower=True)
    fold_test_tokens = extract_unigrams(test_fold, lower=True)

    # Build document-term matrices
    train_dtm, test_dtm = build_term_document_matrix(fold_train_tokens, fold_test_tokens)

    # Convert labels to integers for SVM
    train_labels_int = convert_labels_to_ints(fold_train_labels)  # type:ignore
    test_labels_int = convert_labels_to_ints(fold_test_labels)  # type:ignore

    # Train and evaluate SVM
    svm_classifier = sk.svm.LinearSVC(random_state=42, max_iter=10000)  # type:ignore
    svm_classifier.fit(train_dtm, train_labels_int)
    svm_predictions = svm_classifier.predict(test_dtm)

    # Convert predictions back to string labels for accuracy calculation
    svm_predictions_str = ["POS" if pred == 1 else "NEG" for pred in svm_predictions]
    svm_accuracy = accuracy(svm_predictions_str, fold_test_labels)  # type:ignore
    fold_accuracies_svm.append(svm_accuracy)


# Calculate statistics
mean_accuracy_svm = np.mean(fold_accuracies_svm)
std_accuracy_svm = np.std(fold_accuracies_svm)

print(f"\n{'='*50}")
print(f"10-Fold Cross-Validation Results (Unigrams):")
print(f"SVM Classifier:")
print(f"\tMean Accuracy: {mean_accuracy_svm:.4f}")
print(f"\tStd Deviation: {std_accuracy_svm:.4f}")
print(f"\nNaive Bayes Classifier (baseline):")
print(f"\tMean Accuracy: {mean_accuracy_nb:.4f}")
print(f"\tStd Deviation: {std_accuracy_nb:.4f}")
print(f"{'='*50}")

##################
# raise NotImplementedError

Cross-validation: SVM vs NB: 100%|██████████| 10/10 [00:12<00:00,  1.25s/it]


10-Fold Cross-Validation Results (Unigrams):
SVM Classifier:
	Mean Accuracy: 0.8320
	Std Deviation: 0.0244

Naive Bayes Classifier (baseline):
	Mean Accuracy: 0.8240
	Std Deviation: 0.0282


### POS disambiguation (2pts)

Now add in part-of-speech features. You will find the
movie review dataset has already been POS-tagged for you ([here](https://catalog.ldc.upenn.edu/docs/LDC99T42/tagguid1.pdf) you find the tagset). Try to
replicate the results obtained by Pang et al. (2002).



#### (Q3.2) Replace your features with word+POS features, and report performance with the SVM. Use cross-validation to evaluate the classifier and compare the results with (Q3.1). Does part-of-speech information help? Explain why this may be the case. (1pt)

\####################
\# YOUR ANSWER HERE \#
\####################

Once again, we'll need to adjust our `extract_unigrams` function to extract (token, pos) tuples. Use the provided structure below.

In [ ]:
def extract_pos_unigrams(documents: list[dict], lower: bool = True) -> list[list[tuple[str, str]]]:
    """Extracts (token, pos) tuples.

    Args:
        documents (list[dict]): the list of documents
        lower (bool, optional): whether to lowercase words. Defaults to True.

    Returns:
        list[list[tuple[str, str]]]: the list of (token, pos) tuple lists
    """
    ##################

    all_pos_unigrams = []

    # Iterate over each document
    for doc in documents:
        # Initialize list for (token, pos) tuples in this specific document
        pos_unigrams_per_doc = []

        # Loop over sentences in the document
        for sentence in doc["content"]:
            # Each sentence is already a list of (word, pos) tuples
            for word, pos in sentence:
                # Lowercase the word if requested
                if lower:
                    word = word.lower()

                # Append the (word, pos) tuple
                pos_unigrams_per_doc.append((word, pos))

        # Add this document's tuples to the collection
        all_pos_unigrams.append(pos_unigrams_per_doc)

    return all_pos_unigrams

    ##################
    # raise NotImplementedError

Now train, predict and evaluate your SVM classifier, and compare to the SVM with only unigram features.

In [ ]:
##################

# Store accuracies for SVM with POS features
fold_accuracies_svm_pos = []

# Perform 10-fold cross-validation with word+POS features
for fold_idx, (train_fold, test_fold) in enumerate(
    tqdm.tqdm(cv_splits, desc="Cross-validation: SVM with POS features")
):
    # Extract labels
    fold_train_labels = extract_labels(train_fold)
    fold_test_labels = extract_labels(test_fold)

    # Extract word+POS tuple features
    fold_train_pos_tokens = extract_pos_unigrams(train_fold, lower=True)
    fold_test_pos_tokens = extract_pos_unigrams(test_fold, lower=True)

    # Build document-term matrices
    # Note: build_term_document_matrix can handle tuples as tokens
    train_dtm, test_dtm = build_term_document_matrix(fold_train_pos_tokens, fold_test_pos_tokens)  # type: ignore

    # Convert labels to integers for SVM
    train_labels_int = convert_labels_to_ints(fold_train_labels)  # type: ignore
    test_labels_int = convert_labels_to_ints(fold_test_labels)  # type: ignore

    # Train and evaluate SVM
    svm_classifier_pos = sk.svm.LinearSVC(random_state=42, max_iter=10000)  # type: ignore
    svm_classifier_pos.fit(train_dtm, train_labels_int)
    svm_predictions = svm_classifier_pos.predict(test_dtm)

    # Convert predictions back to string labels for accuracy calculation
    svm_predictions_str = ["POS" if pred == 1 else "NEG" for pred in svm_predictions]
    svm_pos_accuracy = accuracy(svm_predictions_str, fold_test_labels)  # type: ignore
    fold_accuracies_svm_pos.append(svm_pos_accuracy)

# Calculate statistics
mean_accuracy_svm_pos = np.mean(fold_accuracies_svm_pos)
std_accuracy_svm_pos = np.std(fold_accuracies_svm_pos)

print(f"\n{'='*50}")
print(f"10-Fold Cross-Validation Results:")
print(f"\nSVM with Word+POS features:")
print(f"\tMean Accuracy: {mean_accuracy_svm_pos:.4f}")
print(f"\tStd Deviation: {std_accuracy_svm_pos:.4f}")
print(f"\nSVM with Unigrams only (from Q3.1):")
print(f"\tMean Accuracy: {mean_accuracy_svm:.4f}")
print(f"\tStd Deviation: {std_accuracy_svm:.4f}")
print(f"{'='*50}")

##################

# raise NotImplementedError

Cross-validation: SVM with POS features: 100%|██████████| 10/10 [00:22<00:00,  2.23s/it]


10-Fold Cross-Validation Results:

SVM with Word+POS features:
	Mean Accuracy: 0.8380
	Std Deviation: 0.0193

SVM with Unigrams only (from Q3.1):
	Mean Accuracy: 0.8320
	Std Deviation: 0.0244


#### (Q3.3) Discard all closed-class words from your data (keep only nouns, verbs, adjectives, and adverbs), and report performance. Does this help? Use cross-validation to evaluate the classifier and compare the results with (Q3.2). Are closed-class words detrimental to the classifier? Explain why this may be the case. (1pt)

\####################

Removing closed-class words and keeping only open-class words slightly improved classification performance.

By removing closed-class words, we reduce the feature space, which helps prevent overfitting. Closed-class words are very frequent but may not consistently correlate with sentiment. Words like "excellent," "loved," or "disappointed" directly express sentiment, while function words are likely appear with similar frequencies in both positive and negative reviews.

\####################

For the final time, we'll need to adjust our `extract_pos_unigrams` function to filter (token, pos) tuples for open-class words. Use the provided structure below.

In [ ]:
def extract_open_class_pos_unigrams(
    documents: list[dict], lower: bool = True
) -> list[list[tuple[str, str]]]:
    """Extracts (token, pos) tuples for open-class words only.

    Args:
        documents (list[dict]): the list of documents
        lower (bool, optional): whether to lowercase words. Defaults to True.

    Returns:
        list[list[tuple[str, str]]]: the list of (token, pos) tuple lists
    """
    ##################

    # Define open-class POS tags (nouns, verbs, adjectives, adverbs)
    open_class_tags = {
        "NN",
        "NNS",
        "NNP",
        "NNPS",  # Nouns
        "VB",
        "VBD",
        "VBG",
        "VBN",
        "VBP",
        "VBZ",  # Verbs
        "JJ",
        "JJR",
        "JJS",  # Adjectives
        "RB",
        "RBR",
        "RBS",  # Adverbs
    }

    all_open_class_unigrams = []

    # Iterate over each document
    for doc in documents:
        # Initialize list for (token, pos) tuples in this specific document
        open_class_unigrams_per_doc = []

        # Loop over sentences in the document
        for sentence in doc["content"]:
            # Each sentence is a list of (word, pos) tuples
            for word, pos in sentence:
                # Only keep tokens with open-class POS tags
                if pos in open_class_tags:
                    # Lowercase the word if requested
                    if lower:
                        word = word.lower()

                    # Append the (word, pos) tuple
                    open_class_unigrams_per_doc.append((word, pos))

        # Add this document's tuples to the collection
        all_open_class_unigrams.append(open_class_unigrams_per_doc)

    return all_open_class_unigrams

    ##################
    # raise NotImplementedError

Now train, predict and evaluate your SVM classifier, and compare to the SVM with all POS tokens.

In [ ]:
##################

# Store accuracies for SVM with open-class POS features
fold_accuracies_svm_open = []

# Perform 10-fold cross-validation with open-class word+POS features
for fold_idx, (train_fold, test_fold) in enumerate(
    tqdm.tqdm(cv_splits, desc="Cross-validation: SVM with open-class POS features")
):
    # Extract labels
    fold_train_labels = extract_labels(train_fold)
    fold_test_labels = extract_labels(test_fold)

    # Extract open-class word+POS tuple features
    fold_train_open_tokens = extract_open_class_pos_unigrams(train_fold, lower=True)
    fold_test_open_tokens = extract_open_class_pos_unigrams(test_fold, lower=True)

    # Build document-term matrices
    train_dtm, test_dtm = build_term_document_matrix(fold_train_open_tokens, fold_test_open_tokens)  # type: ignore

    # Convert labels to integers for SVM
    train_labels_int = convert_labels_to_ints(fold_train_labels)  # type: ignore
    test_labels_int = convert_labels_to_ints(fold_test_labels)  # type: ignore

    # Train and evaluate SVM
    svm_classifier_open = sk.svm.LinearSVC(random_state=42, max_iter=10000)  # type: ignore
    svm_classifier_open.fit(train_dtm, train_labels_int)
    svm_predictions = svm_classifier_open.predict(test_dtm)

    # Convert predictions back to string labels for accuracy calculation
    svm_predictions_str = ["POS" if pred == 1 else "NEG" for pred in svm_predictions]
    svm_open_accuracy = accuracy(svm_predictions_str, fold_test_labels)  # type: ignore
    fold_accuracies_svm_open.append(svm_open_accuracy)

# Calculate statistics
mean_accuracy_svm_open = np.mean(fold_accuracies_svm_open)
std_accuracy_svm_open = np.std(fold_accuracies_svm_open)

print(f"\n{'='*50}")
print(f"10-Fold Cross-Validation Results:")
print(f"\nSVM with Open-Class Word+POS features:")
print(f"\tMean Accuracy: {mean_accuracy_svm_open:.4f}")
print(f"\tStd Deviation: {std_accuracy_svm_open:.4f}")
print(f"\nSVM with All Word+POS features (from Q3.2):")
print(f"\tMean Accuracy: {mean_accuracy_svm_pos:.4f}")
print(f"\tStd Deviation: {std_accuracy_svm_pos:.4f}")
print(f"{'='*50}")

##################

# raise NotImplementedError

Cross-validation: SVM with open-class POS features: 100%|██████████| 10/10 [00:07<00:00,  1.32it/s]


10-Fold Cross-Validation Results:

SVM with Open-Class Word+POS features:
	Mean Accuracy: 0.8490
	Std Deviation: 0.0189

SVM with All Word+POS features (from Q3.2):
	Mean Accuracy: 0.8380
	Std Deviation: 0.0193


# (4) Discussion (max. 500 words). (5pts)

> Based on your experiments, what are the effective features and techniques in sentiment analysis? What information do different features encode?
Why is this important? What are the limitations of these features and techniques?



*Write your answer here in up to 500 words (-0.25pt for >50 extra words, -0.5 points for >100 extra words, ...)*.


The experiments in this practical reveal several important insights about which features and techniques work best for sentiment analysis in movie reviews. Clear patterns emerged showing what approaches succeed and where traditional NLP classification methods fall short.

Machine learning classifiers (Support Vector Machines and Naive Bayes with smoothing) clearly outperformed lexicon-based approaches. While the lexicon method achieved around 69% accuracy, NB and SVM pushed past 80%. This significant jump shows that statistical patterns learned from training data capture sentiment nuances far better than manually created sentiment dictionaries. The lexicon approach falls short because it depends on fixed word polarities, ignoring context, domain-specific meanings, and the intricate ways people express sentiment in everyday language.

Different feature engineering techniques showed mixed results. Laplace smoothing turned out to be critical for Naive Bayes, solving the zero-probability issue when the model encounters new words during testing. A smoothing factor of 1.0 struck a good balance between generalization and maintaining learned patterns. Stemming cut vocabulary size by 28% by merging morphological variants like “loved” and “loving” into single stems. This dimensionality reduction helped prevent overfitting with only minimal impact on accuracy, indicating that different word forms typically convey similar sentiment.

The n-gram experiments highlighted important compromises. Bigrams and trigrams can capture compositional meaning and things like negation, but they exploded the feature space and created serious data sparsity problems. While feature growth is theoretically exponential with n-gram length, it’s actually slower in practice since many combinations never appear in real text. With limited training data, higher-order n-grams often decreased performance by generating too many rare or one-off features that couldn’t be estimated reliably.
Part-of-speech tagging offered useful disambiguation, helping models differentiate between different uses of identical word forms. 

Filtering down to just open-class words (nouns, verbs, adjectives, adverbs) brought additional modest gains. This approach cuts down noise from common function words that appear at similar rates in both positive and negative reviews, making them weak predictors. Removing them lets the classifier concentrate on features with stronger predictive value.

The biggest limitation across all methods is the bag-of-words representation, which neglects word order beyond local n-grams and misses long-distance dependencies, discourse structure, and nuanced phenomena like sarcasm and irony. Cross-validation consistently demonstrated that matching feature complexity to available training data is essential. Dimensionality remains a core challenge as more sophisticated features demand exponentially more training examples to estimate reliably and avoid overfitting.


# Use of AI tools

By submitting this notebook for grading you testify that:
* AI did not draft an earlier version of your work.
* You did not use AI-powered code completion.
* You did not implement algorithms suggested by an AI tool.
* AI did not revise a version of your work.
* You did not implement suggestions made by an AI tool.

You in the sentences above refers to you and your team member(s). AI refers to LM-based tools and assistants (e.g. ChatGPT, UvA AI Chat, Gemini, etc.).

If you did make use of an AI tool, you should describe the uses you made of it below. Or indicate that no such tool was used.

\####################
\# **YOUR ANSWER HERE** \#
\####################